### ISSUE TITLE
Multiple Output Languages at once (html only for now)

### ISSUE THREAD

**Issue by noah1510** [doxygen/doxygen#8341]
Unlike [Read the docs](https://readthedocs.org/) doxygen is only able to export one language at once instead of all of the languages.
It would be nice if doxygen could export all translations at once (in subfolders of html) and add a language selection in some corner.
It should not be that difficult to implement since there is already support for localization so all that is missing, is just exporting all of them and updating the default template/layout/theme of doxygen.

Below are a few steps that are needed to get this done:

* [ ] export html documentation to `html/lang` instead of `html`
* [ ] add HTML_OUTPUT_LANGUAGES as html option
* [ ] generate the documentation for each language in HTML_OUTPUT_LANGUAGES in `html/lang`
* [ ] update html template with a language selection (similar to [read the docs](https://docs.readthedocs.io/en/stable/localization.html))

For example the English documentation would be in the folder `html/english` or `html/en`, the German documentation in `html/german` or `html/de` and so on.

**Comment by albert-github** [comment]
Interesting idea, though I see some more problems, some thoughts:
- why just limit this to HTML and not also extend to the other output formats by extending `OUTPUT_LANGUAGE` to a list of output languages?
- doxygen does support some localization by means of the `OUTPUT_LANGUAGE` setting but only for the strings specifically generated by doxygen, and it can only have one language active and the structure is not made to handle multiple languages. For this probably the output list generator has to be extended with another layer.
- doxygen is not a translator, so all the user texts have to be translated by the user as well and added to the sources (by means of `\~[language]` commands.
- I foresee some problems in handling of the command `\~[language]` seen the fact that currently only one language can be active (and the texts for the other languages is filtered out).

**Comment by noah1510** [comment]
> why just limit this to HTML and not also extend to the other output formats by extending OUTPUT_LANGUAGE to a list of output languages?

Simply because I have no idea how things like localization works with man pages.
It might work if they are also exported into a `man/lang` subfolder but it might also not work.

> doxygen does support some localization by means of the OUTPUT_LANGUAGE setting but only for the strings specifically generated by doxygen, and it can only have one language active and the structure is not made to handle multiple languages. For this probably the output list generator has to be extended with another layer.

> I foresee some problems in handling of the command \~[language] seen the fact that currently only one language can be active (and the texts for the other languages is filtered out).

To avoid that extending the output generator by another layer could actually help.
My idea is that when this option is active, the OUTPUT_LANGUAGE is always changed to each element in the list and the the output is generated for one language at a time.
Because of that the first step that is needed is to put the output into a language specific subfolder.
So the idea is to change as little as possible and simply extend the output.

> doxygen is not a translator, so all the user texts have to be translated by the user as well and added to the sources (by means of \~[language] commands.

I know and I intended to use exactly that.

By the way is there a way to handle user translations for markdown pages?

**Comment by albert-github** [comment]
> Simply because I have no idea how things like localization works with man pages.

In this case the man pages would land on different directories as well. Should be not that different from normal processing.

> My idea is that when this option is active, the OUTPUT_LANGUAGE is always changed to each element in the list and the the output is generated for one language at a time.

I'm not 100% sure whether this would work but a quick indicative test could be to run the part between
```
  if (Config_getBool(USE_HTAGS))

```
and  (including)
```
  if ( generateHtml &&
       Config_getBool(GENERATE_QHP) &&
      !Config_getString(QHG_LOCATION).isEmpty())
```
multiple times.
First tests would only reset each time the `HTML_OUTPUT` to a different value.
Second tests would not only reset `HTML_OUTPUT` but also the `OUTPUT_LANGUAGE`.
I only think that the extra processing would increase the processing time significantly.

> By the way is there a way to handle user translations for markdown pages?

Not really belonging here, but markdown pages as used in doxygen can use the doxygen commands and thus the `\~...` as well (didn't test it at this moment).

**Comment by noah1510** [comment]
> I only think that the extra processing would increase the processing time significantly.

I also think that the processing time should increase a lot because doxygen has to run for each language.
This could be optimized if all of the selected languages are generated in one pass but I have no idea how the code of doxygen is structured and at the moment I do not have time to look into it.

It seems like it is not as easy as I initially thought it might be...

**Comment by thom-sch** [comment]
A very interesting idea.

But regardless the disadvantages mentioned above:
Using some abilities of Doxygen as there are
- importing additional configuration files
- resolving environment variables

this will work out of the box - s. attached zip-file

[MultiLang.zip](https://github.com/doxygen/doxygen/files/5831284/MultiLang.zip)

Of course this increses the process time; but additionally also increases the disk space and the redunacy if graphs are generated using dot.
Thus the question @doxygen:
Is there a possibility to have general files (such as generated graphics, css and so on)separated in their own directory? This would reduce redundancy dramatically in this case esp. if documentation for many languages is created.

Regards from Meerbusch
Thomas Schneider

**Comment by albert-github** [comment]
Interesting approach for the main page for each language (I assume).
- non invasive in doxygen
- the time problem is of course present as for each language everything has to be generated anew
- I think that even the `@if ...` can be replaced with just `\~...`
- the idea of the `dot` files on a separate directory is interesting, but does, probably, require some small changes as now all `dot` files are in the html directory (and its sub-directories). When in 1 directory the files won't be regenerated as doxygen will find the `.md5` files and conclude no-change so nothing to do. I can think of a directory like `dot` on the html directory, though here will in case of `CREATE_SUBDIRS=YES` also the complete tree of directories be created.

**Comment by noah1510** [comment]
Having the dot files in one directory might also be a problem when dot generates language specific words.
eg. if you have to many members and `UML_LOOK` is set to yes, then dot will add `and 54 more...` when the language is English and `und 54 mehr...` if the language is german.
This text is part of the generated image making it not possible to share the graph between different languages.

**Comment by albert-github** [comment]
@noah1510 Correct, I was afraid of this and in a `svg` file you might probably might be able to manipulate this but in e.g. a `png` file I don't think this is possible. Problem is that the filename of the English and German version will be the same.

**Comment by noah1510** [comment]
Is it possible to detect if dot generates an image with language specific words?
If that is possible, saving those images with different names might be a solution.

At least the call graphs should never have language specific words in them.

**Comment by albert-github** [comment]
@noah1510 the `dot` files are generated by doxygen and here a translator function is called (`tr...`), here there might be some possibilities (probably not easy as the new new has to be transported back to the calling routine and during the generation of the dot file it will be known somewhere during the writing of the dot file that the dot file is language dependent, though I think the later can be overcome "easily" by a simple move of the generated file to another name).
